### ISSUE TITLE
Show sync status - dev

### ISSUE THREAD

**Issue by abbyad** [medic/medic-webapp#1270]
It would be useful for users to know if their data is in sync with the server. 

The first pass at this would be to just display the last sync time to the user, perhaps in the Settings menu. At a later time we can consider also showing a sync status on Reports and Contacts.

**Comment by garethbowen** [comment]
Maybe show a task to get the user to sync if they haven't synced in some time.

**Comment by mandric** [comment]
Might be easy/useful to also include more technical details like the [db info](http://docs.couchdb.org/en/1.6.1/api/database/common.html#get--db) on the about page.

**Comment by garethbowen** [comment]
We have access to the local db info through the [pouch](http://pouchdb.com/api.html#database_information) api. It'd be nice to also grab the remote db info when a connection is available. We might even be able to use this for a replication progress bar.

**Comment by alxndrsn** [comment]
After discussion with @estellecomment, seems logical to have an initial sync when the app is first used.  This will take place behind a splash screen.  Once the initial sync is complete, the app should store a flag locally so that it knows not to do this again.

**Comment by alxndrsn** [comment]
Have made some progress on displaying a splash screen and then moving past it, but this won't be usable until #1780, #1781, #1782 are resolved.  Marking as **`Blocked`**.

**Comment by alxndrsn** [comment]
Splash screen implemented.  Current issues:
- [ ] someone needs to have a look and see if it looks alright
- [ ] there are `alerts` in the code
- [ ] probably safest to get the district and include this in the `query_params.id` before running `.replicate()`

**Comment by SCdF** [comment]
I was hoping that pouchdb replication would provide useful information. So far I've had a super quick look at the event handlers you can add to the replication call: http://pouchdb.com/api.html#replication

Unfortunately they don't seem massively useful by themselves. You effectively have `active`, `paused` and `change` to play with:
 -  `change` fires _when a change has been replicated_.
 - `active` and `pause` are switched between fairly regularly when pouch thinks it might have something useful to do but then realises for whatever reason it doesn't. 
 - `pause` is the end state (since we are continuously replicating) regardless of whether we have changes that need to be duplicated.

Note that there is no obvious indication that you're offline, just that you're not currently actively replicating.

One possibility would be to make sure we funnel all our pouch db put calls through one wrapping function which keeps a list of all the document ids that pass through, and then remove them as they work their way through the change event handler. We could then bind that list (or a count of that list) to the UI and use it to display our sync indication.

Another option would be to just store the last synced seq id somewhere and update it every time the change callback gets hit.

**Comment by mandric** [comment]
These links might be helpful: 
http://stackoverflow.com/questions/29218445/after-successful-replication-are-two-couchdb-databases-guaranteed-to-have-ident
http://stackoverflow.com/questions/29138091/pouchdb-couchdb-sequence-counter-on-replicate

**Comment by estellecomment** [comment]
Separate issue for UX design of sync status : #1849

**Comment by estellecomment** [comment]
Several elements about sync status can be displayed:
 - last sync time
 - number of docs changed locally (from _changes feed, and replication last seq)
 - number of docs changed remotely (if online access, from last seq)

**Comment by estellecomment** [comment]
When online, gives number of docs to send/fetch. Useful for knowing when your replication is done (for now).
No UI changes in this PR, waiting for UX design.

Next steps : 
 - remove the monkeypatch and use a pouch plugin instead
 - expose in UI
 - when offline, get number of changes made locally.

**Comment by SCdF** [comment]
So not having this was truly awful when we were setting up phones today. You have no confidence that you're done. I can understand the argument that in theory you shouldn't need to care, because it will just magically happen in the background. Which may be why it's hard to find out from Pouch.

**Specifically for our use case though**, we don't want to hand a CHP a phone that still has a bunch of forms and other init data to download, because then it will be on their Internet connection, or not at all if the first thing they do is go somewhere with no Internet.

Current best guesses:
 - If you have the phone you can watch the data usage on the about page, but that doesn't tell you much, only that something is happening, but not really what
 - If you have the phone plugged into Chrome development you can watch the network tab
 - If you have the phone plugged into Chrome development you can `location.reload()` and then watch for the log on how long it took to pause `from` replication. If this number is very small (~5s) then this means Pouch asked Couch for docs to replicate, and Couch quickly returned with a batch of 0 because you're done.

This is still pants and we need something better. Tech leads should be able to know with one glance whether or not the sync is done. Stretch goal would be for them to see progress in syncing documents.

**Comment by abbyad** [comment]
I have seen this working, but will defer to @sglangevin in terms of acceptance testing since she has had more experience with it during the training of trainers last week.

**Comment by sglangevin** [comment]
If the goal of this ticket was to display the status of initial replication, then I can confirm that I have seen this working - the status displays and is accurate.

For phone set-up in Uganda (and in Kenya), initial replication fails almost every time because the request times out. I saw 1-2 instances in Uganda where initial replication was successful. Knowing the status of initial replication helps, but because it fails 9/10 times, what would be more helpful would be to have a way to confirm that forms were downloaded. The only way to know this currently is to use Chrome developer tools - on slow connections it's not uncommon for at least some of the forms to fail to download. This poses a risk in projects where the partner is doing phone set-up. This might be a new ticket, but I think we should look into it and see if we can ensure successful phone setup on slower connections (e.g. LG HQ).

**Comment by alxndrsn** [comment]
@sglangevin Right now if you leave the phone alone for a while, the forms should sync (eventually).  If you then restart, you should see the initial replication succeed as there will now be nothing to sync.

One future option is _on first login_ to block the UI until a one-way non-continuous replication success.

**Comment by sglangevin** [comment]
@alxndrsn when you say "leave the phone alone for a while" does this mean that the phone will make multiple tries to download the form definitions? Perhaps I didn't wait long enough, but I saw lots of cases where initial replication failed, then I waited and it tried to download forms during continuous replication, but if they failed, I didn't see it try again. It's possible I needed to wait longer and it would try again, but I didn't see that happen.

**Comment by alxndrsn** [comment]
> I waited and it tried to download forms during continuous replication, but if they failed, I didn't see it try again.

This is very surprising.  After the initial replication, continuous replication starts, which I would expect to retry the downloads periodically until they succeed.

**Comment by sglangevin** [comment]
When you say periodically, does that mean that if they fail, it should try again 5 mins later? Or would it be more like 20 mins later? Perhaps I just didn't wait long enough.

**Comment by alxndrsn** [comment]
Not sure.  From the pouch docs:

> ...will attempt to retry replications in the case of failure (due to being offline), using a backoff algorithm that retries at longer and longer intervals until a connection is re-established.
> – _https://pouchdb.com/api.html#replication_

And from the code:
> ```
> opts.current_back_off = opts.back_off_function(opts.current_back_off);
> ...
> function randomNumber(min, max) {
>   min = parseInt(min, 10) || 0;
>   max = parseInt(max, 10);
>   if (max !== max || max <= min) {
>     max = (min || 1) << 1; //doubling
>   } else {
>     max = max + 1;
>   }
>   var ratio = Math.random();
>   var range = max - min;
>
>   return ~~(range * ratio + min); // ~~ coerces to an int, but fast.
> }
>
> function defaultBackOff(min) {
>   var max = 0;
>   if (!min) {
>     max = 2000;
>   }
>   return randomNumber(min, max);
> }
> ```
> – _https://github.com/pouchdb/pouchdb/blob/master/src/replicate/backoff.js#L17_

Which to me suggests there's no upper limit on the time between retries...so it could take quite a while!!

**Comment by garethbowen** [comment]
> Which to me suggests there's no upper limit on the time between retries...so it could take quite a while!!

Because of this we override the default backoff function with [our own implementation](https://github.com/medic/medic-webapp/blob/develop/static/js/services/db-sync.js#L10) which maxes out at 1 minute between retries.

**Comment by alxndrsn** [comment]
So theoretically the download should retry every minute until it's completed.

**Comment by sglangevin** [comment]
Hmmm. Interesting. I may not have waited long enough and unfortunately I don't have access to slow internet to try again. Did you encounter issues downloading forms when you updated phones in Mukono?

**Comment by sglangevin** [comment]
This feature request turned into the initial replication status being displayed on the About page. That was working well, though has now been removed due to changes in how initial replication is being done. Currently, initial replication status is printed to the console but no longer displayed on the About page because the app does not appear until after initial replication has successfully completed. It retries until completion, then loads the views. Moving to Ready.

Next steps: @diannakane to upload updated mockups to #1849, @garethbowen to consider adding a "last successful sync time" to the 2.7.0 release.
